mantidproject/mantid
# Optimise SANS reductions for event slices

**Description of work.**
This PR introduces new algorithms for the ISIS SANS reduction, where operations which can be performed before event slices are done so. This aims to give significant speed increases for reductions containing many event slices.

A new version of SANSSingleReduction is created. The event slicing logic is moved within this algorithm, and therefore this algorithm outputs group workspaces (instead of matrix workspaces), which contain a workspace for each event slice.

A System test has been created to compare the output from the two versions, and that version 2 runs more quickly than version 1, which was, after all, the point of the new reduction. In testing I have found version 2 to be significantly quicker than version 1, so the likelihood of failures due to random bottlenecks is unlikely.

Because there are a lot of shared functionality between the two versions of the algorithms, I have created a base class from which each version of the algorithm is derived, to reduce code duplication.

The SANS scientists will need to verify this new algorithm, as the outputs are very slightly different. However, once this is done, the old version can be removed.

**To test:**
1. Interfaces -> SANS -> ISIS SANS
2. Load the attached user file and batch file
3. Go to settings tab (on the left)
4. Make sure the reduction mode is front
5. Confirm that **Optimize event slices** is unchecked by default
6. In **Slices**, type `0:300:600`
7. Uncheck **Use compatibility mode**
8. Back in the runs tab, click **Process All**
9. Once that has finished reducing (the row turns green), change the output name in the table to something you would be able to distinguish in the ADS
10. In the logger, find the two logs referring to SANSSingleReduction. Make a note of the combined time they took to run
10. Back in the settings tab, check **Optimize event slices**
11. Back in the runs tab, click **Process All** again. This will run SANSSingleReduction version 2 once (version 2 only needs to be run once for all event slices, whereas version 1 runs once per event slice)
12. In the logger, check how long SANSSingleReduction version 2 took to run. This should be less than the combined time of the two executions of version 1
13. For the two non_trans group workspaces containing event sliced workspaces (i.e. line_1_reduction, and whatever you changed the output name in the table to), plot the `t0.00_T300.00` workspaces on one graph. The two workspaces should look very similar.
14. Do the same with `t300.00_T600.00`

**Data:**
[SANS2DprData.zip](https://github.com/mantidproject/mantid/files/3281787/SANS2DprData.zip)

Fixes #22897 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
